src/diffgraph.jl

```diff
@@ -0,0 +1,141 @@
+
+Struct which holds common sequence of two sequences. Each tuple in `v` contains common elements from first and second
+sequence correspondingly. `l` is the length of the `LCS`, which is constructed as the sum of length of
+common nodes and lengths of wildcard nodes (look also definition of the `lcs` function).
+"""
+struct LCS{T}
+    v::Vector{Tuple{T, T}}
+    l::Int
+end
+
+LCS(t::Tuple{T, T}, l::Int) where T = LCS{T}([t], l)
+Base.:isless(seq1::LCS, seq2::LCS) = seq1.l < seq2.l
+Base.:(*)(seq1::LCS{T}, seq2::LCS{T}) where T = LCS{T}(vcat(seq1.v, seq2.v), seq1.l + seq2.l)
+
+"""
+    lcs(node1::Node, node2::Node, eq::Function, is_wildcard::Function, memo)
+
+For two given nodes, `lcs` greedily constructs longest common sequence of their siblings. Siblings are divided
+into two categories, common nodes which are compared with the help of `eq` function and
+`wildcard` nodes, defined by function `is_wildcard`. `Wildcard` nodes are compared by the length
+of the longest common sequence of their children, so length of the common sequence is defined by
+the sum of the count of equal common nodes and lcs of the wildcard nodes.
+"""
+function lcs(node1::T, node2::T,
+    eq = (x, y) -> x == y, is_wildcard = x -> false,
+    memo = Dict{Tuple{T, T}, LCS{T}}()) where {T <: Node}
+
+    init = (node1, node2)
+    init in keys(memo) && return memo[init]
+
+    if (!is_wildcard(node1) && !is_wildcard(node2) && !eq(node1, node2)) ||
+            (!is_wildcard(node1) && is_wildcard(node2)) ||
+            (is_wildcard(node1) && !is_wildcard(node2))
+
+        seq1 = islastsibling(node2) ? LCS{T}([], 0) : lcs(node1, node2.sibling, eq, is_wildcard, memo)
+        seq2 = islastsibling(node1) ? LCS{T}([], 0) : lcs(node1.sibling, node2, eq, is_wildcard, memo)
+        memo[init] = max(seq1, seq2)
+   elseif !is_wildcard(node1) && !is_wildcard(node2) && eq(node1, node2)
+       if islastsibling(node1) || islastsibling(node2)
+           memo[init] = LCS((node1, node2), 1)
+       else
+           memo[init] = LCS((node1, node2), 1) * lcs(node1.sibling, node2.sibling, eq, is_wildcard, memo)
+       end
+   else
+       seq1 = islastsibling(node2) ? LCS{T}([], 0) : lcs(node1, node2.sibling, eq, is_wildcard, memo)
+       seq2 = islastsibling(node1) ? LCS{T}([], 0) : lcs(node1.sibling, node2, eq, is_wildcard, memo)
+
+       seq3 = if isleaf(node1) || isleaf(node2)
+                  LCS((node1, node2), 1)
+              else
+                  subseq3 = lcs(node1.child, node2.child, eq, is_wildcard, memo)
+                  LCS((node1, node2), subseq3.l + 1)
+              end
+       seq4 = islastsibling(node1) || islastsibling(node2) ? LCS{T}([], 0) : lcs(node1.sibling, node2.sibling, eq, is_wildcard, memo)
+
+       memo[init] = max(seq3*seq4, seq1, seq2)
+   end
+
+   return memo[init]
+end
+
+# sometimes root node has random narrow branches, in order to compare
+# profiles it's easier to remove this spurious branches
+function simplify!(node::Node)
```

Comment:
This is probably too generic a name for an operation that, AFAICT, discards all but the dominant child. Maybe call it `singlechild!`?

Reply:
You are right, but if you do not mind, I'll use name `dominantchild!`, because it shows exactly what I am trying to achieve here.

As a side note, if put it this way, it looks generic enough, that may be at some point it could be ported to `LeftChildRightSiblingTrees`, as a special procedure of sibling pruning. Of course, currently it needs some specific knowledge about nodes in order to compare them.

May be even implement `filter!` function in `LeftChildRightSiblingTrees`? Then here it would like this
```
dominant = maximum(span.(node))
filter!(node, x -> x < dominant)
```

This approach is not needed for this particular problem, I just love high level abstractions :-)

Reply:
`dominantchild!` would be fine.